File: VRTInterstitialCustomEventFyberMarketplace.m
#import "VRTInterstitialCustomEventFyberMarketplace.h"

#import <IASDKCore/IASDKCore.h>

@interface VRTInterstitialCustomEventFyberMarketplace() <IAUnitDelegate, IAMRAIDContentDelegate, IAVideoContentDelegate>
@property (nonatomic, strong) IAAdSpot *adSpot;
@property (nonatomic, strong) IAFullscreenUnitController *unitController;
@property (nonatomic, strong) IAVideoContentController *videoContentController;
@property (nonatomic, strong) IAMRAIDContentController *mraidContentController;
@end

//Fyber Marketplace Interstitial Adapter, Vrtcal as Primary
@implementation VRTInterstitialCustomEventFyberMarketplace

- (void) loadInterstitialAd {
    
    VRTLogWhereAmI();
    
    NSString *spotId = [self.customEventConfig.thirdPartyCustomEventData objectForKey:@"adUnitId"];

    
    IAUserData *userData = [IAUserData build:^(id<IAUserDataBuilder>  _Nonnull builder) {}];
    
    IAAdRequest *adRequest = [IAAdRequest build:^(id<IAAdRequestBuilder>  _Nonnull builder) {
        builder.userData = userData;
        builder.spotID = spotId;
        builder.timeout = 10;
    }];
    
    self.videoContentController = [IAVideoContentController build: ^(id<IAVideoContentControllerBuilder> _Nonnull builder) {
        builder.videoContentDelegate = self;
    }];
    
    self.mraidContentController = [IAMRAIDContentController build: ^(id<IAMRAIDContentControllerBuilder>  _Nonnull builder) {
       builder.MRAIDContentDelegate = self;
    }];

    
    self.unitController = [IAFullscreenUnitController build:^(id<IAFullscreenUnitControllerBuilder>  _Nonnull builder) {
        builder.unitDelegate = self;
        [builder addSupportedContentController:self.mraidContentController];
        [builder addSupportedContentController:self.videoContentController];
    }];
    
    self.adSpot = [IAAdSpot build:^(id<IAAdSpotBuilder>  _Nonnull builder) {
        builder.adRequest = adRequest; // pass here the ad request object;
        // all the supported (by a client side) unit controllers,
        // (in this case - view unit controller) should be added to the desired ad spot:
        [builder addSupportedUnitController:self.unitController];
    }];

    [self.adSpot fetchAdWithCompletion:^(IAAdSpot * _Nullable adSpot, IAAdModel * _Nullable adModel, NSError * _Nullable error) {
        
        if (error) {
            VRTLogWhereAmI();
            [self.customEventLoadDelegate customEventFailedToLoadWithError:error];
            return;
        }

        VRTLogWhereAmI();
        [self.customEventLoadDelegate customEventLoaded];
    }];
}

-(void) showInterstitialAd {
    VRTLogWhereAmI();
    [self.unitController showAdAnimated:YES completion:^{
        VRTLogWhereAmI();
    }];
}

- (UIViewController * _Nonnull)IAParentViewControllerForUnitController:(IAUnitController * _Nullable)unitController {
    VRTLogWhereAmI();
    return [self.viewControllerDelegate vrtViewControllerForModalPresentation];
}




- (void)IAAdDidReceiveClick:(IAUnitController * _Nullable)unitController {
    VRTLogWhereAmI();
    [self.customEventShowDelegate customEventClicked];
}

- (void)IAAdWillLogImpression:(IAUnitController * _Nullable)unitController {
    VRTLogWhereAmI();
    [self.customEventShowDelegate customEventShown];
}

/**
 *  @brief The rewarded units callback for a user reward.
 *
 *  @discussion This callback is called for all type of the rewarded content, both HTML/JS and video (VAST).
 *  In order to use the rewarded callback for all available rewarded content, you will have to implement this method (not the `IAVideoCompleted:`;
 */
- (void)IAAdDidReward:(IAUnitController * _Nullable)unitController {
    VRTLogWhereAmI();
}

- (void)IAUnitControllerWillPresentFullscreen:(IAUnitController * _Nullable)unitController {
    VRTLogWhereAmI();
    [self.customEventShowDelegate customEventWillPresentModal:VRTModalTypeUnknown];
}

- (void)IAUnitControllerDidPresentFullscreen:(IAUnitController * _Nullable)unitController {
    VRTLogWhereAmI();
    [self.customEventShowDelegate customEventDidPresentModal:VRTModalTypeUnknown];
}

- (void)IAUnitControllerWillDismissFullscreen:(IAUnitController * _Nullable)unitController {
    VRTLogWhereAmI();
    [self.customEventShowDelegate customEventWillDismissModal:VRTModalTypeUnknown];
}

- (void)IAUnitControllerDidDismissFullscreen:(IAUnitController * _Nullable)unitController {
    VRTLogWhereAmI();
    [self.customEventShowDelegate customEventDidDismissModal:VRTModalTypeUnknown];
}

- (void)IAUnitControllerWillOpenExternalApp:(IAUnitController * _Nullable)unitController {
    VRTLogWhereAmI();
    [self.customEventShowDelegate customEventWillLeaveApplication];
}

- (void)IAAdDidExpire:(IAUnitController * _Nullable)unitController {
    VRTLogWhereAmI();
}


- (void)IAVideoCompleted:(IAVideoContentController * _Nullable)contentController {
    VRTLogWhereAmI();
}

- (void)IAVideoContentController:(IAVideoContentController * _Nullable)contentController videoInterruptedWithError:(NSError * _Nonnull)error {
    VRTLogWhereAmI();
}

/**
 *  @brief Use to get video duration in seconds. Is valid only if the ad is video ad.
 *
 *  @discussion This method will be invoked after a new received video will become ready to play.
 */
- (void)IAVideoContentController:(IAVideoContentController * _Nullable)contentController videoDurationUpdated:(NSTimeInterval)videoDuration {
    VRTLogWhereAmI();
}

/**
 *  @brief Video progress observer. Use to observe current video progress. Is valid only if the ad is video ad and the video is being played.
 *
 *  @discussion The callback is invoked on the main thread.
 *
 *  @param currentTime Current playback time in seconds.
 *  @param totalTime   Total video duration in seconds.
 */
- (void)IAVideoContentController:(IAVideoContentController * _Nullable)contentController videoProgressUpdatedWithCurrentTime:(NSTimeInterval)currentTime totalTime:(NSTimeInterval)totalTime {
    VRTLogWhereAmI();
}



@end

File: VRTInterstitialCustomEventFyberMarketplace.swift

import IASDKCore
import VrtcalSDK

//Fyber Marketplace Interstitial Adapter, Vrtcal as Primary

class VRTInterstitialCustomEventFyberMarketplace: VRTAbstractInterstitialCustomEvent {
    private var iaAdSpot: IAAdSpot?
    private var iaFullscreenUnitController: IAFullscreenUnitController?
    private var iaVideoContentController: IAVideoContentController?
    private var iaMraidContentController: IAMRAIDContentController?
    
    private var iaUnitDelegatePassthrough = IAUnitDelegatePassthrough()
    private var iaVideoContentDelegatePassthrough = IAVideoContentDelegatePassthrough()
    private var iaMRAIDContentDelegatePassthrough = IAMRAIDContentDelegatePassthrough()
    
    override func loadInterstitialAd() {
        
        VRTLogInfo()
        
        
        guard let spotId = customEventConfig.thirdPartyAdUnitId(
            customEventLoadDelegate: customEventLoadDelegate
        ) else {
            return
        }
        
        // Set Delegates
        iaUnitDelegatePassthrough.viewControllerDelegate = viewControllerDelegate
        iaMRAIDContentDelegatePassthrough.customEventShowDelegate = customEventShowDelegate
        
        
        guard let iaAdRequest = IAAdRequest.build({ builder in
            builder.spotID = spotId
            builder.timeout = 10
        }) else {
            let vrtError = VRTError(
                vrtErrorCode: .customEvent,
                message: "Could not make iaAdRequest"
            )
            customEventLoadDelegate?.customEventFailedToLoad(vrtError: vrtError)
            return
        }
        
        // videoContentController
        guard let videoContentController = IAVideoContentController.build({ builder in
            builder.videoContentDelegate = self.iaVideoContentDelegatePassthrough
        }) else {
            
            let vrtError = VRTError(
                vrtErrorCode: .customEvent,
                message: "Could not make iaViewUnitController"
            )
            customEventLoadDelegate?.customEventFailedToLoad(vrtError: vrtError)
            return
        }
        
        // mraidContentController
        guard let mraidContentController = IAMRAIDContentController.build({ builder in
            builder.mraidContentDelegate = self.iaMRAIDContentDelegatePassthrough
        }) else {
            let vrtError = VRTError(
                vrtErrorCode: .customEvent,
                message: "Could not make mraidContentController"
            )
            customEventLoadDelegate?.customEventFailedToLoad(vrtError: vrtError)
            return
        }
        
        // iaFullscreenUnitController
        guard let iaFullscreenUnitController = IAFullscreenUnitController.build({ builder in
            builder.unitDelegate = self.iaUnitDelegatePassthrough
            builder.addSupportedContentController(mraidContentController)
            builder.addSupportedContentController(videoContentController)
        }) else {
            let vrtError = VRTError(
                vrtErrorCode: .customEvent,
                message: "Could not make iaFullscreenUnitController"
            )
            customEventLoadDelegate?.customEventFailedToLoad(vrtError: vrtError)
            return
        }
        self.iaFullscreenUnitController = iaFullscreenUnitController
        
        // iaAdSpot
        guard let iaAdSpot = IAAdSpot.build({ builder in
            builder.adRequest = iaAdRequest // pass here the ad request object;
            // all the supported (by a client side) unit controllers,
            // (in this case - view unit controller) should be added to the desired ad spot:
            builder.addSupportedUnitController(iaFullscreenUnitController)
        }) else {
            let vrtError = VRTError(
                vrtErrorCode: .customEvent,
                message: "Could not make iaAdSpot"
            )
            customEventLoadDelegate?.customEventFailedToLoad(vrtError: vrtError)
            return
        }
        
        // fetchAd
        iaAdSpot.fetchAd() { adSpot, adModel, error in
            if let error {
                VRTLogInfo()
                let vrtError = VRTError(
                    vrtErrorCode: .customEvent,
                    error: error
                )
                self.customEventLoadDelegate?.customEventFailedToLoad(vrtError: vrtError)
                return
            }
            
            VRTLogInfo()
            self.customEventLoadDelegate?.customEventLoaded()
        }
    }
    
    override func showInterstitialAd() {
        VRTLogInfo()
        iaFullscreenUnitController?.showAd(animated: true) {
            VRTLogInfo()
        }
    }
}
    
    






/*
IAMRAIDContentDelegate, IAVideoContentDelegate {



    func iaAdDidReceiveClick(_ unitController: IAUnitController?) {
        VRTLogInfo()
        customEventShowDelegate?.customEventClicked()
    }

    func iaAdWillLogImpression(_ unitController: IAUnitController?) {
        VRTLogInfo()
        customEventShowDelegate?.customEventShown()
    }

    ///  The rewarded units callback for a user reward.
    ///
    ///  - Remark: This callback is called for all type of the rewarded content, both HTML/JS and video (VAST).
    ///  In order to use the rewarded callback for all available rewarded content, you will have to implement this method (not the `IAVideoCompleted:`;
    func iaAdDidReward(_ unitController: IAUnitController?) {
        VRTLogInfo()
    }

    func iaUnitControllerWillPresentFullscreen(_ unitController: IAUnitController?) {
        VRTLogInfo()
        customEventShowDelegate?.customEventWillPresentModal(.unknown)
    }

    func iaUnitControllerDidPresentFullscreen(_ unitController: IAUnitController?) {
        VRTLogInfo()
        customEventShowDelegate?.customEventDidPresentModal(.unknown)
    }

    func iaUnitControllerWillDismissFullscreen(_ unitController: IAUnitController?) {
        VRTLogInfo()
        customEventShowDelegate?.customEventWillDismissModal(.unknown)
    }

    func iaUnitControllerDidDismissFullscreen(_ unitController: IAUnitController?) {
        VRTLogInfo()
        customEventShowDelegate?.customEventDidDismissModal(.unknown)
    }

    func iaUnitControllerWillOpenExternalApp(_ unitController: IAUnitController?) {
        VRTLogInfo()
        customEventShowDelegate?.customEventWillLeaveApplication()
    }

    func iaAdDidExpire(_ unitController: IAUnitController?) {
        VRTLogInfo()
    }

    func iaVideoCompleted(_ contentController: IAVideoContentController?) {
        VRTLogInfo()
    }

    func iaVideoContentController(_ contentController: IAVideoContentController?, videoInterruptedWithError error: Error) {
        VRTLogInfo()
    }

    ///  Use to get video duration in seconds. Is valid only if the ad is video ad.
    ///
    ///  - Remark: This method will be invoked after a new received video will become ready to play.
    func iaVideoContentController(_ contentController: IAVideoContentController?, videoDurationUpdated videoDuration: TimeInterval) {
        VRTLogInfo()
    }

    ///  Video progress observer. Use to observe current video progress. Is valid only if the ad is video ad and the video is being played.
    ///
    ///  - Remark: The callback is invoked on the main thread.
    ///
    ///  - Parameters:
    ///   - currentTime: Current playback time in seconds.
    ///    - totalTime:   Total video duration in seconds.
    func iaVideoContentController(_ contentController: IAVideoContentController?, videoProgressUpdatedWithCurrentTime currentTime: TimeInterval, totalTime: TimeInterval) {
        VRTLogInfo()
    }
}


*/
